pkg/model/message.go

```diff
@@ -14,6 +15,83 @@ type Messages struct {
 	Original bool
 }
 
+type MessagesSlice []Messages
+
+// HasLanguage checks if MessagesSlice contains Messages with the given language.
+func (ms MessagesSlice) HasLanguage(lang language.Tag) bool {
+	return ms.LanguageIndex(lang) != -1
+}
+
+// LanguageIndex returns index of Messages with the given language. If not found, returns -1.
+func (ms MessagesSlice) LanguageIndex(lang language.Tag) int {
+	return slices.IndexFunc(ms, func(m Messages) bool {
+		return m.Language == lang
+	})
+}
+
+// Replace replaces Messages with the same language. If not found, appends to the slice.
+func (ms *MessagesSlice) Replace(messages Messages) {
+	switch idx := ms.LanguageIndex(messages.Language); idx {
+	case -1:
+		*ms = append(*ms, messages)
+	default:
+		(*ms)[idx] = messages
+	}
+}
+
+// SplitOriginal returns a pointer to the original and other messages.
+func (m MessagesSlice) SplitOriginal() (original *Messages, others MessagesSlice) {
+	others = make(MessagesSlice, 0, len(m))
+
+	for i := range m {
+		if m[i].Original {
+			original = &m[i]
+		} else {
+			others = append(others, m[i])
+		}
+	}
+
+	return
+}
+
+/*
+AlterTranslations changes the status of specific messages in all languages to UNTRANSLATED
+based on a list of message IDs, while keeping the original language messages status.
+
+Example:
+
+	untranslatedIDs := { "1" }
+
+	{ Language: en, Original: true, Messages: [ { ID: "1", Message: "Hello", Status: Translated  } ], ...
+	{ Language: fr, Messages: [ { ID: "1", Message: "Bonjour", Status: Translated  } ], ... ] }
+	{ Language: de, Messages: [ { ID: "1", Message: "Hallo", Status: Translated  } ], ... ]
+
+	Result:
+	{ Language: en, Original: true, Messages: [ { ID: "1", Message: "Hello", Status: Translated  } ], ...
+	{ Language: fr, Messages: [ { ID: "1", Message: "Bonjour", Status: Untranslated  }, ... ] }
+	{ Language: de, Messages: [ { ID: "1", Message: "Hallo", Status: Untranslated  } ], ... ]
+*/
+func (ms MessagesSlice) AlterTranslations(untranslatedIDs []string) {
```

Comment:
What principle do you apply when choosing what functions to implement in the `model` package? `AlterTranslations` is here, but `populateTranslations` and `getUntranslatedIDs` are in the `server` package. What are their scope and intent?

Reply:
Did not want to make extra changes in this PR, as it is already quite large.

But conceptually `populateTranslations` should go to `model` pkg with the receiver `MessagesSlice`. It seems a good fit, by looking at function parameters.

`getUntranslatedIDs` I guess should remain in `server` pkg. As it's scope is limited.

Reply:
It's worth to make an extra effort. The basic principle we can apply is to move everything to the `model` package that does data transformation - single responsibility.

Also, `AlterTranslations` name is too abstract and does not tell what it does.

```go
// in the handler
all.MarkUntranslated(original.FindUntranslatedIDs(previous))
```